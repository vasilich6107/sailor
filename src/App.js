import React, { Component } from 'react';
import logo from './assets/images/logo.png';
import './App.css';

const APPLICANT_STATES = {
  APPLIED: 'APPLIED',
  INTERVIEWING: 'INTERVIEWING',
  HIRED: 'HIRED',
};

const DASHBOARD_COLUMNS = [APPLICANT_STATES.APPLIED, APPLICANT_STATES.INTERVIEWING, APPLICANT_STATES.HIRED];

class App extends Component {
  state = {
    filter: {
      name: localStorage.getItem('nameFilter') || null,
      city: localStorage.getItem('cityFilter') || null,
    },
    applicants: {
      loading: false,
      data: null,
      error: null,
    },
  };

  componentDidMount() {
    this.setState(previousState => ({
      applicants: {
        ...previousState.applicants,
        loading: true,
      },
    }));

    fetch('https://randomuser.me/api/?nat=gb&results=5')
      .then(response => response.json())
      .then(({ results }) => {
        const data = results.reduce((acc, applicant) => {
          acc[applicant.login.uuid] = {
            ...applicant,
            state: APPLICANT_STATES.APPLIED,
          };

          return acc;
        }, {});

        this.setState(previousState => ({
          applicants: {
            ...previousState.applicants,
            loading: false,
            data,
          },
        }));
      })
      .catch(error => {
        this.setState(previousState => ({
          applicants: {
            ...previousState.applicants,
            loading: false,
            error,
          },
        }));
      });
  }

  getDashboardApplicants = () => {
    const applicants = Object.values(this.state.applicants.data);

    return applicants.reduce(
      (acc, applicant) => {
        if (
          (this.state.filter.name === null && this.state.filter.city === null) ||
          (this.state.filter.name && applicant.name.first.indexOf(this.state.filter.name) !== -1) ||
          (this.state.filter.name && applicant.name.last.indexOf(this.state.filter.name) !== -1) ||
          (this.state.filter.city && applicant.location.city.indexOf(this.state.filter.city) !== -1)
        ) {
          acc[applicant.state].push(applicant);
        }

        return acc;
      },
      {
        [APPLICANT_STATES.APPLIED]: [],
        [APPLICANT_STATES.INTERVIEWING]: [],
        [APPLICANT_STATES.HIRED]: [],
      }
    );
  };

  handleApplicantStateChange = (uuid, newState) => {
    this.setState(prevState => {
      const data = {
        ...prevState.applicants.data,
        [uuid]: {
          ...prevState.applicants.data[uuid],
          state: newState,
        },
      };

      return {
        applicants: {
          ...prevState.applicants,
          data,
        },
      };
    });
  };

  render() {
    if (this.state.applicants.data) {
      const dashboardApplicants = this.getDashboardApplicants();
      console.log(dashboardApplicants);

      return (
        <div className="App">
          <header className="App-header">
            <img src={logo} className="App-logo" alt="logo" />
            <h1 className="App-title">OpenOceanStudio: Crew Applications</h1>
          </header>
          <div className="filter">
            <div>Name</div>
            <div>
              <input
                type="text"
                value={this.state.filter.name}
                onChange={({ target: { value } }) => {
                  localStorage.setItem('nameFilter', value);

                  return this.setState(prevState => ({
                    filter: {
                      ...prevState.filter,
                      name: value || null,
                    },
                  }));
                }}
              />
            </div>
            <div>Country</div>
            <div>
              <input
                type="text"
                value={this.state.filter.city}
                onChange={({ target: { value } }) => {
                  localStorage.setItem('cityFilter', value);

                  return this.setState(prevState => ({
                    filter: {
                      ...prevState.filter,
                      city: value || null,
                    },
                  }));
                }}
              />
            </div>
          </div>
          <div className="dashboard">
            {DASHBOARD_COLUMNS.map((column, index) => (
              <div className="dashboard__column" key={column}>
                <div>{column}</div>
                {dashboardApplicants[column].map(applicant => (
                  <div key={applicant.login.uuid}>
                    <div>
                      <div>
                        applicant {applicant.name.first} {applicant.name.last} {applicant.location.city}
                      </div>
                    </div>
                    <div>
                      {DASHBOARD_COLUMNS[index - 1] && (
                        <button
                          type="button"
                          onClick={() => {
                            this.handleApplicantStateChange(applicant.login.uuid, DASHBOARD_COLUMNS[index - 1]);
                          }}
                        >
                          rev
                        </button>
                      )}

                      {DASHBOARD_COLUMNS[index + 1] && (
                        <button
                          type="button"
                          onClick={() => {
                            this.handleApplicantStateChange(applicant.login.uuid, DASHBOARD_COLUMNS[index + 1]);
                          }}
                        >
                          ff
                        </button>
                      )}
                    </div>
                  </div>
                ))}
              </div>
            ))}
          </div>
        </div>
      );
    }

    return null;
  }
}

export default App;
