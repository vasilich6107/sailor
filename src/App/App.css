body {
  margin: 0;
  padding: 0;
  font-family: sans-serif;
}

.App {
  text-align: center;
}

.App-logo {
  height: 80px;
}

.App-header {
  background-color: #06679a;
  height: 150px;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}
