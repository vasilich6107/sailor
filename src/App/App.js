import React from 'react';
import logo from '../assets/images/logo.png';
import Dashboard from '../Dashboard';

import './App.css';

const App = () => (
  <div className="App">
    <header className="App-header">
      <img src={logo} className="App-logo" alt="logo" />
      <h1 className="App-title">OpenOceanStudio: Crew Applications</h1>
    </header>
    <Dashboard />
  </div>
);

export default App;
