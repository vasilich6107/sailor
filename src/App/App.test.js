import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

it('Renders main app with dashboard', () => {
  expect(shallow(<App />)).toMatchSnapshot();
});
