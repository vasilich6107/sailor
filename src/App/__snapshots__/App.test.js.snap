// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders main app with dashboard 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.png"
    />
    <h1
      className="App-title"
    >
      OpenOceanStudio: Crew Applications
    </h1>
  </header>
  <Dashboard />
</div>
`;
